### PR TITLE
firefox: fix cross

### PR DESCRIPTION
**Pull request by jnbr**
mostly @Gottox's [work](https://gist.githubusercontent.com/Gottox/7c6931e9a53291a91972d8b25b4badac/raw/e676748a26afab76ee21b674aa4f13dca6c57e2f/gistfile1.txt)

builds tested for:
- [x] aarch64
- [x] aarch64-musl
- [x] armv7l
- [x] armv7l-musl
- [ ] ~armv6l~ won't happen anytime soon
- [ ] ~armv6l-musl~ won't happen anytime soon
- [x] ppc64le
- [x] ppc64le-musl

### CODE REVIEW

**Comment by jnbr** [comment]
`armv7l` fails with `echo: write error`

**Comment by jnbr** [comment]
`armv6` and `armv7l-musl` fail with:
```
21:14.64 In file included from /builddir/firefox-64.0/third_party/aom/aom_dsp/arm/blend_a64_mask_neon.c:12:
21:14.64 /usr/lib/gcc/armv7l-linux-musleabihf/8.2.0/include/arm_neon.h:10303:1: error: inlining failed in call to always_inline 'vld1_u8': target specific option mismatch
21:14.64  vld1_u8 (const uint8_t * __a)
21:14.64  ^~~~~~~
```

**Comment by jnbr** [comment]
current status:

- aarch64 (glibc and musl) build completed
- armv6hf (glibc and musl) fail in trying to build rust crates for armv7:
```
20:43.79    Compiling nodrop v0.1.12
20:43.82 error[E0463]: can't find crate for `std`
20:43.82   |
20:43.82   = note: the `armv7-unknown-linux-gnueabihf` target may not be installed
20:43.82 error: aborting due to previous error
20:43.82 For more information about this error, try `rustc --explain E0463`.
20:43.84 error: Could not compile `nodrop`.
20:43.84 To learn more, run the command again with --verbose.
20:43.84 make[4]: *** [/builddir/firefox-64.0/config/rules.mk:1046: force-cargo-library-build] Error 101
20:43.84 make[3]: *** [/builddir/firefox-64.0/config/recurse.mk:74: toolkit/library/rust/target] Error 2
20:43.84 make[2]: *** [/builddir/firefox-64.0/config/recurse.mk:34: compile] Error 2
20:43.84 make[1]: *** [/builddir/firefox-64.0/config/rules.mk:431: default] Error 2
20:43.84 make: *** [client.mk:125: build] Error 2
```

- armv7l (glibc andd musl) still fail with `echo: write error` (glibc) and `echo:` (musl)

**Comment by jnbr** [comment]
slightly more useful error from armv7l-musl
```
172:27.72 {standard input}: Assembler messages:
172:27.72 {standard input}:1033: Error: bad expression -- `b #_ZL14SigprofHandleriP9siginfo_tPv'
```

**Comment by jnbr** [comment]
armv7 builds but without webrtc for now.
armv6 is still broken because of the use of NEON extensions in libaom

**Comment by Gottox** [comment]
What was the write error with armv7?

File: srcpkgs/firefox/patches/fix-tools.patch [hunk 1]
```diff
@@ -39,4 +51,4 @@
 +#endif
```
Comment (Gottox) [inline srcpkgs/firefox/patches/fix-tools.patch]: where's the #if for that #endif?

**Comment by jnbr** [comment]
> What was the write error with armv7?

multiple definitions and undefined references:
```
31:01.31 /usr/lib/gcc/armv7l-linux-gnueabihf/8.2.0/../../../../armv7l-linux-gnueabihf/bin/ld: ../../media/webrtc/trunk/webrtc/modules/audio_coding/isac_fix_c_gn/lattice_armv7.o: in function `WebRtcIsacfix_FilterA
rLoop':
31:01.31 /builddir/firefox-64.0/media/webrtc/trunk/webrtc/modules/audio_coding/isac_fix_c_gn/../codecs/isac/fix/source/lattice_armv7.S:34: multiple definition of `WebRtcIsacfix_FilterArLoop'; ../../media/webrtc/t
runk/webrtc/modules/audio_coding/isac_fix_c_gn/Unified_c_isac_fix_c_gn1.o:/builddir/firefox-64.0/media/webrtc/trunk/webrtc/modules/audio_coding/codecs/isac/fix/source/lattice_c.c:30: first defined here
31:01.31 /usr/lib/gcc/armv7l-linux-gnueabihf/8.2.0/../../../../armv7l-linux-gnueabihf/bin/ld: ../../media/webrtc/trunk/webrtc/modules/audio_coding/isac_fix_c_gn/pitch_filter_armv6.o: in function `WebRtcIsacfix_Pi
tchFilterCore':
31:01.31 /builddir/firefox-64.0/media/webrtc/trunk/webrtc/modules/audio_coding/isac_fix_c_gn/../codecs/isac/fix/source/pitch_filter_armv6.S:33: multiple definition of `WebRtcIsacfix_PitchFilterCore'; ../../media/
webrtc/trunk/webrtc/modules/audio_coding/isac_fix_c_gn/pitch_filter_c.o:/builddir/firefox-64.0/media/webrtc/trunk/webrtc/modules/audio_coding/codecs/isac/fix/source/pitch_filter_c.c:28: first defined here
31:01.31 /usr/lib/gcc/armv7l-linux-gnueabihf/8.2.0/../../../../armv7l-linux-gnueabihf/bin/ld: ../../media/webrtc/trunk/webrtc/system_wrappers/cpu_features_linux_gn/Unified_c_pu_features_linux_gn0.o: in function `
WebRtc_GetCPUFeaturesARM':
31:01.31 /builddir/firefox-64.0/media/webrtc/trunk/webrtc/system_wrappers/source/cpu_features_linux.c:29: multiple definition of `WebRtc_GetCPUFeaturesARM'; ../../media/webrtc/trunk/webrtc/system_wrappers/system_
wrappers_gn/Unified_cpp_system_wrappers_gn0.o:/builddir/firefox-64.0/media/webrtc/trunk/webrtc/system_wrappers/source/cpu_features.cc:78: first defined here
31:01.31 /usr/lib/gcc/armv7l-linux-gnueabihf/8.2.0/../../../../armv7l-linux-gnueabihf/bin/ld: ../../media/webrtc/trunk/webrtc/common_audio/common_audio_c_gn/Unified_c_common_audio_c_gn1.o: in function `InitFuncti
onPointers':
31:01.31 /builddir/firefox-64.0/media/webrtc/trunk/webrtc/common_audio/signal_processing/spl_init.c:93: undefined reference to `WebRtcSpl_MaxAbsValueW16Neon'
31:01.31 /usr/lib/gcc/armv7l-linux-gnueabihf/8.2.0/../../../../armv7l-linux-gnueabihf/bin/ld: /builddir/firefox-64.0/media/webrtc/trunk/webrtc/common_audio/signal_processing/spl_init.c:93: undefined reference to
`WebRtcSpl_MaxAbsValueW32Neon'
31:01.31 /usr/lib/gcc/armv7l-linux-gnueabihf/8.2.0/../../../../armv7l-linux-gnueabihf/bin/ld: /builddir/firefox-64.0/media/webrtc/trunk/webrtc/common_audio/signal_processing/spl_init.c:93: undefined reference to
`WebRtcSpl_MaxValueW16Neon'
31:01.31 /usr/lib/gcc/armv7l-linux-gnueabihf/8.2.0/../../../../armv7l-linux-gnueabihf/bin/ld: /builddir/firefox-64.0/media/webrtc/trunk/webrtc/common_audio/signal_processing/spl_init.c:93: undefined reference to
`WebRtcSpl_MaxValueW32Neon'
31:01.31 /usr/lib/gcc/armv7l-linux-gnueabihf/8.2.0/../../../../armv7l-linux-gnueabihf/bin/ld: /builddir/firefox-64.0/media/webrtc/trunk/webrtc/common_audio/signal_processing/spl_init.c:93: undefined reference to
`WebRtcSpl_MinValueW16Neon'
31:01.31 /usr/lib/gcc/armv7l-linux-gnueabihf/8.2.0/../../../../armv7l-linux-gnueabihf/bin/ld: /builddir/firefox-64.0/media/webrtc/trunk/webrtc/common_audio/signal_processing/spl_init.c:93: undefined reference to
`WebRtcSpl_MinValueW32Neon'
31:01.31 /usr/lib/gcc/armv7l-linux-gnueabihf/8.2.0/../../../../armv7l-linux-gnueabihf/bin/ld: /builddir/firefox-64.0/media/webrtc/trunk/webrtc/common_audio/signal_processing/spl_init.c:93: undefined reference to
`WebRtcSpl_CrossCorrelationNeon'
31:01.31 /usr/lib/gcc/armv7l-linux-gnueabihf/8.2.0/../../../../armv7l-linux-gnueabihf/bin/ld: /builddir/firefox-64.0/media/webrtc/trunk/webrtc/common_audio/signal_processing/spl_init.c:93: undefined reference to
`WebRtcSpl_DownsampleFastNeon'
31:01.31 /usr/lib/gcc/armv7l-linux-gnueabihf/8.2.0/../../../../armv7l-linux-gnueabihf/bin/ld: ../../media/webrtc/trunk/webrtc/common_audio/common_audio_gn/Unified_cpp_common_audio_gn0.o: in function `webrtc::FIRF
ilter::Create(float const*, unsigned int, unsigned int)':
31:01.31 /builddir/firefox-64.0/media/webrtc/trunk/webrtc/common_audio/fir_filter.cc:63: undefined reference to `webrtc::FIRFilterNEON::FIRFilterNEON(float const*, unsigned int, unsigned int)'
31:01.31 /usr/lib/gcc/armv7l-linux-gnueabihf/8.2.0/../../../../armv7l-linux-gnueabihf/bin/ld: ../../media/webrtc/trunk/webrtc/common_audio/common_audio_gn/Unified_cpp_common_audio_gn0.o: in function `webrtc::Sinc
Resampler::Resample(unsigned int, float*)':
31:01.31 /builddir/firefox-64.0/media/webrtc/trunk/webrtc/common_audio/resampler/sinc_resampler.cc:316: undefined reference to `webrtc::SincResampler::Convolve_NEON(float const*, float const*, float const*, doubl
e)'
31:01.32 /usr/lib/gcc/armv7l-linux-gnueabihf/8.2.0/../../../../armv7l-linux-gnueabihf/bin/ld: ../../media/webrtc/trunk/webrtc/modules/audio_processing/audio_processing_c_gn/Unified_c_udio_processing_c_gn0.o: in function `WebRtcNsx_InitCore':
31:01.32 /builddir/firefox-64.0/media/webrtc/trunk/webrtc/modules/audio_processing/ns/nsx_core.c:633: undefined reference to `WebRtcNsx_NoiseEstimationNeon'
31:01.32 /usr/lib/gcc/armv7l-linux-gnueabihf/8.2.0/../../../../armv7l-linux-gnueabihf/bin/ld: /builddir/firefox-64.0/media/webrtc/trunk/webrtc/modules/audio_processing/ns/nsx_core.c:633: undefined reference to `WebRtcNsx_PrepareSpectrumNeon'
31:01.32 /usr/lib/gcc/armv7l-linux-gnueabihf/8.2.0/../../../../armv7l-linux-gnueabihf/bin/ld: /builddir/firefox-64.0/media/webrtc/trunk/webrtc/modules/audio_processing/ns/nsx_core.c:633: undefined reference to `WebRtcNsx_SynthesisUpdateNeon'
31:01.32 /usr/lib/gcc/armv7l-linux-gnueabihf/8.2.0/../../../../armv7l-linux-gnueabihf/bin/ld: /builddir/firefox-64.0/media/webrtc/trunk/webrtc/modules/audio_processing/ns/nsx_core.c:633: undefined reference to `WebRtcNsx_AnalysisUpdateNeon'
31:01.32 /usr/lib/gcc/armv7l-linux-gnueabihf/8.2.0/../../../../armv7l-linux-gnueabihf/bin/ld: ../../media/webrtc/trunk/webrtc/modules/audio_processing/audio_processing_c_gn/Unified_c_udio_processing_c_gn0.o: in function `WebRtcNsx_ComputeSpectralFlatness':
31:01.32 /builddir/firefox-64.0/media/webrtc/trunk/webrtc/modules/audio_processing/ns/nsx_core.c:1049: undefined reference to `WebRtcNsx_kLogTableFrac'
31:01.32 /usr/lib/gcc/armv7l-linux-gnueabihf/8.2.0/../../../../armv7l-linux-gnueabihf/bin/ld: /builddir/firefox-64.0/media/webrtc/trunk/webrtc/modules/audio_processing/ns/nsx_core.c:1049: undefined reference to `WebRtcNsx_kLogTableFrac'
```

File: srcpkgs/firefox/patches/fix-tools.patch (hunk 1 above)
Comment (jnbr) [reply srcpkgs/firefox/patches/fix-tools.patch]: 4 lines above

**Comment by jnbr** [comment]
> What was the write error with armv7?

it came from this:

```
172:27.72 {standard input}: Assembler messages:
172:27.72 {standard input}:1033: Error: bad expression -- `b #_ZL14SigprofHandleriP9siginfo_tPv'
```

fixed by not doing the `MOZ_SIGNAL_TRAMPOLINE` https://github.com/void-linux/void-packages/pull/6660/files#diff-e8d34830e5a650ba7970444c447a1e20R35

adopted the fix from this patch: https://salsa.debian.org/mozilla-team/firefox/blob/release/master/debian/patches/fixes/Bug-1463035-Remove-MOZ_SIGNAL_TRAMPOLINE.-r-darchons.patch

File: srcpkgs/firefox/patches/fix-tools.patch (hunk 1 above)
Comment (Gottox) [reply srcpkgs/firefox/patches/fix-tools.patch]: Ah, the Github view had hid it.

**Comment by jnbr** [comment]
`armv7` builds break with `XBPS_CCACHE` and `XBPS_MAKEJOBS > 1` can someone reproduce?

**Comment by jnbr** [comment]
mentioned ccache problem was worked around by adding `-Wno-psabi` to `CFLAGS`

**Comment by jnbr** [comment]
`ppc64le` builds, `ppc64le-musl` fails because of a bug in the vendored `libc` rust crate. This is fixed [upstream](https://github.com/rust-lang/libc/commit/6d9c681f4efd8670bb0606d8ad4da32bcf799a32) but I haven't figured out yet how to apply patches to vendored rust dependencies.

**Comment by Gottox** [comment]
May this help? https://doc.rust-lang.org/cargo/reference/manifest.html#the-patch-section

**Comment by Gottox** [comment]
> `armv7` builds break with `XBPS_CCACHE` and `XBPS_MAKEJOBS > 1` can someone reproduce?

Does cleaning the ccache help?

**Comment by jnbr** [comment]
> Does cleaning the ccache help?

no, happens with empty ccache as well

**Comment by jnbr** [comment]
fixed webrtc for armv7

**Comment by jnbr** [comment]
> May this help? https://doc.rust-lang.org/cargo/reference/manifest.html#the-patch-section

Yes, thanks. Didn't work for me before because I tried to patch to the wrong version.

**Pull request merged**